editor/test/internal/cache_test.clj

```diff
@@ -57,16 +54,171 @@
-                                      (gt/endpoint 1 :c) 3})
-                      (cache-hit [(gt/endpoint 1 :a)
-                                  (gt/endpoint 1 :c)])
-                      (cache-encache {(gt/endpoint 1 :d) 4}))]
-       (are [k v] (= v (get snapshot k))
+(deftest limits-test
+  (letfn [(populate-cache [cache]
+            (-> cache
+              (cache-encache {(gt/endpoint 1 :a) 1
+                              (gt/endpoint 1 :b) 2
+                              (gt/endpoint 1 :c) 3})
+              (cache-hit [(gt/endpoint 1 :a)
+                          (gt/endpoint 1 :c)])
+              (cache-encache {(gt/endpoint 1 :d) 4})))]
+
+    (testing "positive cache limit"
+      (with-clean-system {:cache-size 3}
+        (let [snapshot (populate-cache cache)]
+          (is (= 3 (count snapshot)))
+          (are [k v] (= v (get snapshot k))
             (gt/endpoint 1 :a) 1
             (gt/endpoint 1 :c) 3
             (gt/endpoint 1 :d) 4))))
+
+    (testing "zero cache limit (no caching)"
+      (with-clean-system {:cache-size 0}
+        (let [snapshot (populate-cache cache)]
+          (is (empty? snapshot)))))
+
+    (testing "unlimited cache"
+      (with-clean-system {:cache-size -1}
+        (let [snapshot (populate-cache cache)]
+          (is (= 4 (count snapshot)))
+          (are [k v] (= v (get snapshot k))
+            (gt/endpoint 1 :a) 1
+            (gt/endpoint 1 :b) 2
+            (gt/endpoint 1 :c) 3
+            (gt/endpoint 1 :d) 4))))))
+
+(deftest clear-test
+  (letfn [(populate-cache [cache]
+            (cache-encache cache {(gt/endpoint 1 :a) 1
+                                  (gt/endpoint 1 :b) 2
+                                  (gt/endpoint 1 :c) 3
+                                  (gt/endpoint 1 :d) 4}))]
+
+    (testing "positive cache limit"
+      (with-clean-system {:cache-size 3}
+        (let [cache (populate-cache cache)]
+          (is (= 3 (count cache)))
+          (let [cache (cache-clear cache)]
+            (is (= 0 (count cache)))
+            (let [cache (populate-cache cache)]
+              (is (= 3 (count cache))))))))
+
+    (testing "zero cache limit (no caching)"
+      (with-clean-system {:cache-size 0}
+        (is (identical? null-cache cache))
+        (let [cache (populate-cache cache)]
+          (is (= 0 (count cache)))
+          (let [cache (cache-clear cache)]
+            (is (= 0 (count cache)))
+            (is (identical? null-cache cache))
+            (let [cache (populate-cache cache)]
+              (is (= 0 (count cache))))))))
+
+    (testing "unlimited cache"
+      (with-clean-system {:cache-size -1}
+        (is (unlimited-cache? cache))
+        (let [cache (populate-cache cache)]
+          (is (= 4 (count cache)))
+          (let [cache (cache-clear cache)]
+            (is (= 0 (count cache)))
+            (is (unlimited-cache? cache))
+            (let [cache (populate-cache cache)]
+              (is (= 4 (count cache))))))))))
```

Comment:
Note that `"retain rules remain after clear"` is tested in the `retain-test` below